File: macos/Sources/JingClawKit/Reduce.swift
import Foundation

/// What a client puts where folded turns were.
public let foldNotice = "[earlier turns were folded into a summary]"

/// Folds an event log into what a client shows.
///
/// The same reducer as the console's and the daemon's, in a third language,
/// and checked against the same cases. Nothing else stops three
/// implementations drifting apart, and a drift here means somebody watching
/// the same session in two places is told two different things about what the
/// agent did.
public enum Reducer {
    /// Applies one event.
    public static func reduce(_ state: SessionState, _ event: LogEvent) -> SessionState {
        var next = state
        next.headSeq = event.seq

        switch event.kind {
        case "user.message":
            next.messages.append(Message(role: .user, text: event.body.text ?? ""))

        case "assistant.delta":
            // Joined onto the open assistant turn. A delta that starts a new
            // message is one word on a line of its own.
            let index = openAssistant(next.messages) ?? {
                next.messages.append(Message(role: .assistant))
                return next.messages.count - 1
            }()
            next.messages[index].text += event.body.text ?? ""

        case "tool.requested":
            // Attached to the turn that asked, which is the assistant turn
            // being written — creating one if the model asked for a tool
            // before saying anything.
            let index = openAssistant(next.messages) ?? {
                next.messages.append(Message(role: .assistant))
                return next.messages.count - 1
            }()
            next.messages[index].toolCalls.append(ToolCall(name: event.body.name ?? ""))

        case "tool.completed":
            markCompleted(
                &next.messages,
                name: event.body.name ?? "",
                failed: event.body.isError ?? false
            )

        case "approval.requested":
            next.pendingApprovals.append(event.body.approvalID ?? "")

        case "approval.resolved":
            let resolved = event.body.approvalID ?? ""
            next.pendingApprovals.removeAll { $0 == resolved }

        case "conversation.compacted":
            // Everything before the fold is a summary now, so what a client
            // draws is the notice and whatever follows.
            next.messages = [Message(role: .assistant, text: foldNotice)]

        case "run.state_changed":
            let run = event.body.runID ?? ""
            let status = event.body.status ?? ""
            if ["completed", "failed", "cancelled"].contains(status) {
                if next.activeRun == run { next.activeRun = "" }
            } else {
                next.activeRun = run
            }

        default:
            break
        }

        return next
    }

    /// Folds a whole sequence, which is what a client does on attach.
    public static func reduceAll(_ events: [LogEvent]) -> SessionState {
        events.reduce(SessionState()) { reduce($0, $1) }
    }

    /// The assistant turn being written, or nil.
    ///
    /// The last message when it is the assistant's. A user turn closes it:
    /// what follows belongs to the answer to that turn, not to the one before.
    private static func openAssistant(_ messages: [Message]) -> Int? {
        guard let last = messages.indices.last, messages[last].role == .assistant else {
            return nil
        }
        return last
    }

    /// Finishes the last call of that name still running.
    ///
    /// The last rather than the first: names repeat within a turn, and marking
    /// the first reports the wrong one as done.
    private static func markCompleted(_ messages: inout [Message], name: String, failed: Bool) {
        for messageIndex in messages.indices.reversed() {
            for callIndex in messages[messageIndex].toolCalls.indices.reversed() {
                let call = messages[messageIndex].toolCalls[callIndex]
                if call.name == name && !call.completed {
                    messages[messageIndex].toolCalls[callIndex].completed = true
                    messages[messageIndex].toolCalls[callIndex].isError = failed
                    return
                }
            }
        }
    }
}

File: core/internal/webui/assets/reduce.js
// The reducer three clients share.
//
// Turning an event log into a screen is done once per client, in a different
// language each time, and nothing stops them drifting except a set of examples
// they are all checked against. Those live in fixtures/session-view.json; this
// is the implementation they check.
//
// Pure on purpose. The console's rendering used to fold events straight into
// the DOM, which meant the behaviour could not be tested without a browser and
// could not be compared with Swift's at all.

export const ROLE_USER = 'user';
export const ROLE_ASSISTANT = 'assistant';

export const FOLD_NOTICE = '[earlier turns were folded into a summary]';

export function emptyState() {
  return { messages: [], pending_approvals: [], active_run: '', head_seq: 0 };
}

// reduce applies one event and returns the new state.
export function reduce(state, event) {
  const next = {
    messages: state.messages.map(cloneMessage),
    pending_approvals: [...state.pending_approvals],
    active_run: state.active_run,
    head_seq: Number(event.seq),
  };

  const body = event.body || {};

  switch (event.kind) {
    case 'user.message':
      next.messages.push({ role: ROLE_USER, text: body.text || '' });
      break;

    case 'assistant.delta': {
      // Joined onto the open assistant turn. A delta that starts a new message
      // is one word on a line of its own.
      const index = openAssistant(next.messages);
      if (index < 0) next.messages.push({ role: ROLE_ASSISTANT, text: '' });
      const at = index < 0 ? next.messages.length - 1 : index;
      next.messages[at].text += body.text || '';
      break;
    }

    case 'tool.requested': {
      // Attached to the turn that asked, which is the assistant turn being
      // written — creating one if the model asked before saying anything.
      const index = openAssistant(next.messages);
      if (index < 0) next.messages.push({ role: ROLE_ASSISTANT, text: '' });
      const at = index < 0 ? next.messages.length - 1 : index;
      const message = next.messages[at];
      message.tool_calls = message.tool_calls || [];
      message.tool_calls.push({
        name: body.name || '', completed: false, is_error: false,
      });
      break;
    }

    case 'tool.completed':
      markCompleted(next.messages, body.name || '', Boolean(body.is_error));
      break;

    case 'approval.requested':
      next.pending_approvals.push(body.approval_id || '');
      break;

    case 'approval.resolved':
      next.pending_approvals =
        next.pending_approvals.filter((id) => id !== body.approval_id);
      break;

    case 'conversation.compacted':
      // Everything before the fold is a summary now, so what a client draws is
      // the notice and whatever follows.
      next.messages = [{ role: ROLE_ASSISTANT, text: FOLD_NOTICE }];
      break;

    case 'run.state_changed':
      if (['completed', 'failed', 'cancelled'].includes(body.status)) {
        if (next.active_run === body.run_id) next.active_run = '';
      } else {
        next.active_run = body.run_id || '';
      }
      break;

    default:
      break;
  }

  return next;
}

export function reduceAll(events) {
  return events.reduce(reduce, emptyState());
}

function cloneMessage(message) {
  const copy = { role: message.role, text: message.text };
  if (message.tool_calls) copy.tool_calls = message.tool_calls.map((c) => ({ ...c }));
  return copy;
}

// openAssistant is the assistant turn being written, or -1.
//
// The last message when it is the assistant's. A user turn closes it: what
// follows belongs to the answer to that turn, not to the one before it.
function openAssistant(messages) {
  if (messages.length === 0) return -1;
  const last = messages.length - 1;
  return messages[last].role === ROLE_ASSISTANT ? last : -1;
}

// markCompleted finishes the last call of that name still running.
//
// The last rather than the first: names repeat within a turn, and marking the
// first reports the wrong one as done.
function markCompleted(messages, name, failed) {
  for (let i = messages.length - 1; i >= 0; i -= 1) {
    const calls = messages[i].tool_calls || [];
    for (let j = calls.length - 1; j >= 0; j -= 1) {
      if (calls[j].name === name && !calls[j].completed) {
        calls[j].completed = true;
        calls[j].is_error = failed;
        return;
      }
    }
  }
}
